swap/src/common/mod.rs

```diff
@@ -0,0 +1,203 @@
+pub mod tracing_util;
+
+use std::{collections::HashMap, path::PathBuf};
+
+use anyhow::anyhow;
+use tokio::{fs::{read_dir, File}, io::{AsyncBufReadExt, BufReader}};
+use uuid::Uuid;
+
+const LATEST_RELEASE_URL: &str = "https://github.com/comit-network/xmr-btc-swap/releases/latest";
+
+#[derive(Clone, Debug, PartialEq, Eq)]
+pub enum Version {
+    Current,
+    Available,
+}
+
+/// Check the latest release from GitHub API.
+pub async fn check_latest_version(current_version: &str) -> anyhow::Result<Version> {
+    let response = reqwest::get(LATEST_RELEASE_URL).await?;
+    let e = "Failed to get latest release.";
+    let download_url = response.url();
+    let segments = download_url.path_segments().ok_or_else(|| anyhow!(e))?;
+    let latest_version = segments.last().ok_or_else(|| anyhow!(e))?;
+
+    let result = if is_latest_version(current_version, latest_version) {
+        Version::Current
+    } else {
+        tracing::warn!(%current_version, %latest_version, %download_url,
+            "You are not on the latest version",
+        );
+        Version::Available
+    };
+
+    Ok(result)
+}
+
+// todo: naive implementation can be improved using semver
+fn is_latest_version(current: &str, latest: &str) -> bool {
+    current == latest
+}
+
+/// helper macro for [`redact`]... eldrich sorcery
+/// the macro does in essence the following:
+/// 1. create a static regex automaton for the pattern
+/// 2. find all matching patterns using regex
+/// 3. create a placeholder for each distinct matching pattern
+/// 4. add the placeholder to the hashmap
+macro_rules! regex_find_placeholders {
+    ($pattern:expr, $create_placeholder:expr, $replacements:expr, $input:expr) => {{
+        // compile the regex pattern
+        static REGEX: once_cell::sync::Lazy<regex::Regex> = once_cell::sync::Lazy::new(|| {
+            tracing::debug!("initializing regex");
+            regex::Regex::new($pattern).expect("invalid regex pattern")
+        });
+
+        // keep count of count patterns to generate distinct placeholders
+        let mut counter: usize = 0;
+
+        // for every matched address check whether we already found it
+        // and if we didn't, generate a placeholder for it
+        for address in REGEX.find_iter($input) {
+            if !$replacements.contains_key(address.as_str()) {
+                $replacements.insert(address.as_str().to_owned(), $create_placeholder(counter));
+                counter += 1;
+            }
+        }
```

Comment:
This doesn't work unless we run the function on the entire file at once. The counter wont be preserved across logs. Different addresses might get the same placeholder

Reply:
I added a function `redact_with` that takes `& mut HashMap<...>` to hoist the state. currently testing